### PR TITLE
fix(dart_frog_cli): report daemon missing parameters

### LINKED ISSUE #913: fix(dart_frog_cli): daemon should report when there is missing params
**Description**

Running the daemon and sending a request with a missing parameters field returns an error with a message that could be more descriptive.

**Steps To Reproduce**

1. Create a Dart Frog project:

```sh
dart_frog create sample
```

2. Start the daemon on the project (from /sample):
```sh
dart_frog daemon
```

3. Send an invalid message without params (within the deamon session):
```
[{"id":"1","method":"dev_server.start"}]
```

4. The error message is:
```
[{"id":"1","error":{"message":"Malformed message, invalid workingDirectory"}}]
```

**Expected Behavior**

To report that the request is missing a prams field with the required parameters instead of reporting an invalid workingDirectory.

**Screenshots**

Reproductive steps output from terminal:
CleanShot 2023-08-09 at 08 57 26

**Additional Context**

 - https://github.com/VeryGoodOpenSource/dart_frog/blob/88a65d939964b0c2786e771850ce4a3cfc637871/packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart#L42

### PR DESCRIPTION
**Pull request by PiotrFLEURY**
## Status

**READY**

## Description

Fix #913 

`dart_frog_cli` Daemon wil now report missing parameters for `dev_server` and `route_configuration` domains

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/dart_frog_cli/lib/src/daemon/protocol.dart [hunk 1]
```diff
@@ -118,6 +118,38 @@ class DaemonRequest extends DaemonMessage {
...
+  ///   `final paramValue = getParam<String>('requiredValue')`.
+  ///
+  /// Optional parameters should be typed as nullable
+  ///  `final paramValue = getParam<String?>('optionalValue')`.
+  ///
+  T getParam<T>(String name) {
+    final param = params?[name];
+    if (param is! T) {
+      // Check if param type is optional or not
+      if (param == null && <T?>[] is! List<T>) {
```
Comment (renancaraujo) [inline packages/dart_frog_cli/lib/src/daemon/protocol.dart:141]:
```suggestion
      if (param == null && null is! T) {
```


Just an alternative, feel free to ignore :D

File: packages/dart_frog_cli/lib/src/daemon/protocol.dart [hunk 2]
```diff
@@ -118,6 +118,38 @@ class DaemonRequest extends DaemonMessage {
...
+  /// is not of type [T].
+  ///
+  /// Required parameters should be typed as not nullable
+  ///   `final paramValue = getParam<String>('requiredValue')`.
+  ///
+  /// Optional parameters should be typed as nullable
+  ///  `final paramValue = getParam<String?>('optionalValue')`.
+  ///
+  T getParam<T>(String name) {
+    final param = params?[name];
```
Comment (renancaraujo) [inline packages/dart_frog_cli/lib/src/daemon/protocol.dart:138]:
```suggestion
    final params = this.params;
    
    if(params == null) {
      throw DartFrogDaemonMalformedMessageException(
        'Missing params object',
      );
    }

    final param = params[name];
```

**Review by renancaraujo — COMMENTED** [review]
Can we have a test on `protocol_test` for at least the case of missing `params` ?

Everything else:

looks very good to me

File: packages/dart_frog_cli/lib/src/daemon/protocol.dart (hunk 1 above)
Comment (PiotrFLEURY) [reply packages/dart_frog_cli/lib/src/daemon/protocol.dart]: Sounds better to understand

Done

File: packages/dart_frog_cli/lib/src/daemon/protocol.dart (hunk 2 above)
Comment (PiotrFLEURY) [reply packages/dart_frog_cli/lib/src/daemon/protocol.dart]: Done

**Comment by PiotrFLEURY** [comment]
Thanks for the review @renancaraujo 

I just pushed the suggested modification with some bonus tests for the getParam method ✨

Feel free to ask for any new change

**Review by erickzanardo — APPROVED** [review]

**Review by renancaraujo — APPROVED** [review]

File: packages/dart_frog_cli/lib/src/daemon/exceptions.dart [hunk 3]
```diff
@@ -27,3 +27,14 @@ class DartFrogDaemonMalformedMessageException
   const DartFrogDaemonMalformedMessageException(String message)
       : super('Malformed message, $message');
 }
+
+/// {@template dart_frog_daemon_missing_parameter_exception}
+/// An exception thrown when the daemon reports
+/// a missing parameter of a message.
```
Comment (alestiago) [inline packages/dart_frog_cli/lib/src/daemon/exceptions.dart:33]: Nit: force doc to reach max length
```suggestion
/// An exception thrown when the daemon reports a missing parameter of a
/// message.
```

**Review by alestiago — APPROVED** [review]

**Review by renancaraujo — APPROVED** [review]

**Review by alestiago — APPROVED** [review]

**Review by erickzanardo — APPROVED** [review]

**Pull request merged**